src/value_generators/src/DocumentGenerator.cpp

```diff
@@ -1388,6 +1391,57 @@ class ArrayGenerator : public Generator<bsoncxx::array::value> {
     const UniqueGenerator<int64_t> _nTimesGen;
 };
 
+/** `{^DistinctArray: {of: {^RandomInt {start: 10, end: 1000}}}, number: 2}` */
+template <typename T>
+class DistinctArrayGenerator : public Generator<bsoncxx::array::value> {
+public:
+    DistinctArrayGenerator(const Node& node, GeneratorArgs generatorArgs)
+        : _rng{generatorArgs.rng},
+          _node{node},
+          _generatorArgs{generatorArgs},
+          _valueGen{getValueGen(node, generatorArgs)},
+          _nTimesGen{intGenerator(extract(node, "number", "^DistinctArray"), generatorArgs)} {}
+
+    bsoncxx::array::value evaluate() override {
+        bsoncxx::builder::basic::array builder{};
+        auto times = _nTimesGen->evaluate();
+        // Choose initial bucket size of 2N to minimize hash collisions and resizes.
+        std::unordered_set<T> distinctValues(2 * times);
+        while (distinctValues.size() < times) {
```

Comment:
This could infinite-loop on mis-configs. E.g. 

    {^DistinctArray: {of: {^RandomInt {start: 10, end: 10}}}, number: 10}

Hard to detect in the general case from config alone, but maybe stop the iteration and barf after like 100*N iterations or something?